File: services/eval/src/metrics/recall.ts
export function recallAtK(retrievedIds: string[], relevantIds: string[], k: number): number {
  if (k <= 0) throw new Error(`k must be positive, got ${k}`);
  if (relevantIds.length === 0) return 0;

  const relevantSet = new Set(relevantIds);
  const topK = retrievedIds.slice(0, k);
  const hits = topK.reduce((acc, id) => acc + (relevantSet.has(id) ? 1 : 0), 0);
  return hits / relevantSet.size;
}

File: services/eval/src/rag_eval/metrics/recall.py
from collections.abc import Sequence


def recall_at_k(retrieved_ids: Sequence[str], relevant_ids: Sequence[str], k: int) -> float:
    """Fraction of all (deduplicated) relevant ids recovered within the top-k."""
    if k <= 0:
        raise ValueError(f"k must be positive, got {k}")
    if len(relevant_ids) == 0:
        return 0.0

    relevant_set = set(relevant_ids)
    top_k = retrieved_ids[:k]
    hits = sum(1 for chunk_id in top_k if chunk_id in relevant_set)
    return hits / len(relevant_set)
